### PR TITLE
GSFA and iGSFA nodes

### PR DESCRIPTION
**Pull request by AlbertoEsc**
Dear MDP developers and enthusiasts,

I am glad to share the following code for your consideration and possible inclusion in the MDP-toolkit. It consists of two new nodes that implement graph-based SFA (GSFA) and information-preserving GSFA (iGSFA). 

GSFA is a generalization of SFA that has been designed for supervised dimensionality reduction and facilitates the solution of regression and classification problems, whereas iGSFA enhances the feature representation of (G)SFA. The improved feature representation is particularly useful when the algorithm is applied hierarchically by means of hierarchical iGSFA (HiGSFA) networks. 

Why should we include these algorithms? HiGSFA has consistently yielded slower top level features and better generalization than HSFA. For example, for age estimation from facial images (MORPH-II database) HiGSFA yields a mean absolute error of 3.41 years, which was state-of-the-art accuracy for some months. Also gender and ethnicity estimation were improved.
On a more standard database, MNIST, HiGSFA yields a classification error of 0.67% (preprocessing: 2-pixel image shifts, hyperparameters computed on hold out validation sets). As far as I know this is better than previous versions of SFA and HSFA. Face localization and detection also benefit from HiGSFA. 
Even unsupervised learning applications benefit from the new extensions. An experiment where the input data is the simulated view of a rat moving inside a box shows that HiSFA (a special case of HiGSFA) clearly improves top level feature slowness compared to HSFA. For this experiment is HiSFA about 8 times more data/sample efficient than HSFA.

I will highly appreciate all your feedback regarding the code, style, tests, and documentation. Currently all test pass using pytest 3.5.0 but some fail using 3.4.0 and older versions. What is the oldest version of pytest that we should support? 

Thank you very much for your time and consideration,

Alberto Escalante
PS 1. HiGSFA has been explained in this manuscript: https://arxiv.org/abs/1601.03945. However, this version has become obsolete by now, if you would like to get an updated version just write me a short e-mail.

PS 2. pytest 3.5 gives hundreds of warnings “Metafunc.addcall is deprecated and scheduled to be removed in pytest 4.0. Please use Metafunc.parametrize instead.”, perhaps we should consider updating this part of the tests.

### CODE REVIEW

File: mdp/test/test_nodes_generic.py [hunk 1]
```diff
@@ -282,6 +282,9 @@ def NeuralGasNode_inp_arg_gen():
 def LinearRegressionNode_inp_arg_gen():
     return uniform(size=(1000, 5))
 
+def iGSFANode_inp_arg_gen():
+    return uniform(size=(1000, 4)) * 0.01  #
```
Comment (Stewori) [inline mdp/test/test_nodes_generic.py]: Why the trailing '#'?

File: mdp/test/test_nodes_generic.py [hunk 2]
```diff
@@ -345,6 +348,13 @@ def CCIPCANode_inp_arg_gen():
          init_args=[(nodes.PolynomialExpansionNode, (1,), {}),
                     (nodes.PolynomialExpansionNode, (1,), {}),
                     True]),
+    dict(klass='iGSFANode',
+         init_args=[None, None, None, None, None, None, None, 0.5, False, False],
+#         """pre_expansion_node_class=None, pre_expansion_out_dim=None,
```
Comment (Stewori) [inline mdp/test/test_nodes_generic.py]: With '#', """ can be cleaned out

File: mdp/nodes/gsfa_nodes.py [hunk 3]
```diff
@@ -0,0 +1,2040 @@
...
+from __future__ import division
+
+import scipy
+import scipy.optimize
+import sys
+
+# NodeException, TrainingException, ValueError
+import mdp
+from mdp import numx, NodeException, TrainingException
+# from mdp.nodes.sfa_nodes import SFANode
```
Comment (Stewori) [inline mdp/nodes/gsfa_nodes.py]: I'm not sure whether it is official MDP styleguide, but usually out-commented code should be avoided.
I'm personally fine with out-commented code if it serves some documentary purpose, e.g. document good-looking yet subtly wrong approach, too naive approach, a pitfall, something a future maintainer might be intrigued to break by an attempt to "simplify".
In that case it should however be accompanied by an explaining comment.

There are some more out-commented code snippets below. Please check those accordingly.

**Comment by Stewori** [comment]
Hey, I gave some minor cosmetic comments. Apart from that it looks good to me, but I don't actually have time for a deep review. If tests pass and the others have no concerns I'd trust your expertiese and would recommend that we can include these algorithms. Thanks a lot for your work!

Re PS2: Please open a separate issue for that.
Re PS1: Please make sure that link is included in the doc-comment of the according nodes (maybe it already is, I didn't read in detail.) If you have an updated version, maybe it can be made publicly available to be linked or included in the documentation?

**Comment by AlbertoEsc** [comment]
Thanks a lot for your feedback Stefan. Your modifications are easy to apply, but I will wait a bit for further comments before updating the pull request

**Comment by AlbertoEsc** [comment]
Dear MDP developers,
the pull request has been updated. The main changes are the following: 1) I applied Stefan's suggestions, 2) I tried to make the code more consistent with existing mdp code (e.g., removed use of pytest.approx, reduced margin to 79 characters, marked functions that require scipy), and 3) I cleaned the code a bit, e.g., considering suggestions by pylint.
Let me know if you have further feedback,
Alberto

**Comment by Stewori** [comment]
I opened PS 2 as #39 to keep track. Will now commit and close this PR.
